src/Microsoft.ML.Data/Data/SchemaDefinition.cs

```diff
@@ -392,37 +392,42 @@ public static SchemaDefinition Create(Type userType, Direction direction = Direc
 
                 InternalSchemaDefinition.GetVectorAndItemType(memberInfo, out bool isVector, out Type dataType);
 
-                PrimitiveDataViewType itemType;
-                var keyAttr = memberInfo.GetCustomAttribute<KeyTypeAttribute>();
-                if (keyAttr != null)
-                {
-                    if (!KeyDataViewType.IsValidDataType(dataType))
-                        throw Contracts.ExceptParam(nameof(userType), "Member {0} marked with KeyType attribute, but does not appear to be a valid kind of data for a key type", memberInfo.Name);
-                    if (keyAttr.KeyCount == null)
-                        itemType = new KeyDataViewType(dataType, dataType.ToMaxInt());
-                    else
-                        itemType = new KeyDataViewType(dataType, keyAttr.KeyCount.Count.GetValueOrDefault());
-                }
-                else
-                    itemType = ColumnTypeExtensions.PrimitiveTypeFromType(dataType);
-
                 // Get the column type.
                 DataViewType columnType;
-                var vectorAttr = memberInfo.GetCustomAttribute<VectorTypeAttribute>();
-                if (vectorAttr != null && !isVector)
-                    throw Contracts.ExceptParam(nameof(userType), $"Member {memberInfo.Name} marked with {nameof(VectorTypeAttribute)}, but does not appear to be a vector type", memberInfo.Name);
-                if (isVector)
+                if (TypeManager.GetDataViewTypeOrNull(dataType) == null)
```

Comment:
>GetDataViewTypeOrNull [](start = 32, length = 21)

this usage pattern implies an overload like:
bool TryGetDataViewType(out datatype) overload.

Reply:
+1

---
In reply to: [284945327](https://github.com/dotnet/machinelearning/pull/3263#discussion_r284945327) [](ancestors = 284945327)

Reply:
alternatively, you can assign on line 369 and remove the else condition on line 430.
columnType = TypeManager.GetDataViewTypeOrNull(dataType);

---
In reply to: [285245746](https://github.com/dotnet/machinelearning/pull/3263#discussion_r285245746) [](ancestors = 285245746,284945327)

Reply:
Just add `Knows(...)`.

---
In reply to: [285260595](https://github.com/dotnet/machinelearning/pull/3263#discussion_r285260595) [](ancestors = 285260595,285245746,284945327)